REST API: Create Metrics endpoint

tinahollygb opened the pull request:
The payload tries to adhere to `ApiMetric` as closely as possible while allowing everything to be optional. This includes some conversion math, as well as remapping of field names.

It also un-nests `sql.builder` to `sqlBuilder`.

It includes 2 main functions:

- a validation function that returns a validation result
- a transformation function that transforms the generated payload to a `MetricInterface`
 
The above 2 points are probably most important, though it may also make sense to test the code generation I added/changed.


## Testing Requests & Responses

Test making requests and responses.

<details>
<summary><b>Orders metric in the API</b></summary>

request:

```json
{
  "datasourceId": "ds_iwuvkhpc0dkv",
  "type": "count",
  "name": "Orders metric in the API",
  "sql": {
    "identifierTypes": ["user_id"],
    "conversionSQL": "SELECT userid as user_id, timestamp as timestamp, 1 as value FROM sample.orders"
  },
  "behavior": {
    "goal": "increase",
    "conversionWindowStart": 168,
    "conversionWindowEnd": 240,
    "minPercentChange": 0,
    "minSampleSize": 10
  }
}
```

response:

```json
{
  "metric": {
    "id": "met_sktwi229elfu9nr1o",
    "name": "Orders metric in the API",
    "description": "",
    "dateCreated": "2023-03-29T22:36:31.020Z",
    "dateUpdated": "2023-03-29T22:36:31.020Z",
    "archived": false,
    "datasourceId": "ds_iwuvkhpc0dkv",
    "owner": "",
    "projects": [],
    "tags": [],
    "type": "count",
    "behavior": {
      "goal": "increase",
      "cap": 0,
      "minPercentChange": 0,
      "maxPercentChange": 0.6,
      "minSampleSize": 10,
      "riskThresholdDanger": 0.0125,
      "riskThresholdSuccess": 0.0025,
      "conversionWindowStart": 168,
      "conversionWindowEnd": 240
    },
    "sql": {
      "identifierTypes": [
        "user_id"
      ],
      "conversionSQL": "SELECT userid as user_id, timestamp as timestamp, 1 as value FROM sample.orders",
      "userAggregationSQL": "SUM(value)",
      "denominatorMetricId": ""
    }
  }
}

```

</details>



<details>
<summary><b>average order value</b></summary>

Request:

```json
{
  "datasourceId": "ds_iwuvkhpc0dkv",
  "type": "revenue",
  "name": "Average Order Value metric in the API",
  "owner": "luke",
  "sql": {
    "identifierTypes": ["user_id"],
    "denominatorMetricId": "met_sktwi229elfu9nr1o",
    "conversionSQL": "SELECT userid as user_id, timestamp as timestamp, amount as value FROM sample.orders"
  },
  "behavior": {
    "goal": "increase",
    "cap": 1200,
    "conversionWindowStart": 0,
    "conversionWindowEnd": 168,
    "minPercentChange": 0,
    "minSampleSize": 10
  }
}
```

Response:

```json
{
  "metric": {
    "id": "met_sktwi229elfu9qzpv",
    "name": "Average Order Value metric in the API",
    "description": "",
    "dateCreated": "2023-03-29T22:39:02.227Z",
    "dateUpdated": "2023-03-29T22:39:02.227Z",
    "archived": false,
    "datasourceId": "ds_iwuvkhpc0dkv",
    "owner": "luke",
    "projects": [],
    "tags": [],
    "type": "revenue",
    "behavior": {
      "goal": "increase",
      "cap": 1200,
      "minPercentChange": 0,
      "maxPercentChange": 0.6,
      "minSampleSize": 10,
      "riskThresholdDanger": 0.0125,
      "riskThresholdSuccess": 0.0025,
      "conversionWindowStart": 0,
      "conversionWindowEnd": 168
    },
    "sql": {
      "identifierTypes": [
        "user_id"
      ],
      "conversionSQL": "SELECT userid as user_id, timestamp as timestamp, amount as value FROM sample.orders",
      "userAggregationSQL": "SUM(value)",
      "denominatorMetricId": "met_sktwi229elfu9nr1o"
    }
  }
}
```

</details>

tinahollygb commented on packages/back-end/src/services/experiments.ts:
```diff
@@ -813,6 +816,115 @@ export function toSnapshotApiInterface(
+    projects,
+    inverse: behavior?.goal === "decrease",
+    ignoreNulls: false,
+    aggregation: sql?.userAggregationSQL || mixpanel?.userAggregation || "",
+    sql: sql?.userAggregationSQL || sql?.conversionSQL, // TODO: Is this correct?
+    cap: behavior?.cap ?? 0,
+    conditions: conditionsForDataSourceType(),
+
+    conversionDelayHours: behavior?.conversionWindowStart ?? 0,
+    conversionWindowHours: behavior?.conversionWindowEnd ?? 0,
+    loseRisk: behavior?.riskThresholdDanger ?? 0.0125,
+    winRisk: behavior?.riskThresholdSuccess ?? 0.0025,
+    column: sqlBuilder?.valueColumnName || mixpanel?.eventValue, // Is this correct?
+    maxPercentChange:
+      behavior?.maxPercentChange ?? metricDefaults?.maxPercentageChange ?? 0.5,
+    minPercentChange:
+      behavior?.minPercentChange ??
+      metricDefaults?.minPercentageChange ??
+      0.005,
+    minSampleSize:
+      behavior?.minSampleSize ?? metricDefaults?.minimumSampleSize ?? 150,
+    denominator: sql?.denominatorMetricId || "",
+    table: mixpanel?.eventName || sqlBuilder?.tableName || "",
+    status: "active",
+    timestampColumn: sqlBuilder?.timestampColumnName || "timestamp",
+    queries: [],
+    userIdTypes: sql?.identifierTypes || [],
+    userIdType: undefined,
+    anonymousIdColumn: "", // TODO: ??
+    userIdColumn: "", // TODO: Remove?
+    userIdColumns: (sqlBuilder?.identifierTypeColumns || []).reduce<
+      Record<string, string>
+    >((acc, { columnName, identifierType }) => {
+      acc[columnName] = identifierType;
+      return acc;
+    }, {}),
+  };
+
+  return metric;
+}
```
Looking for feedback for this util, which is called from the `postMetric` handler.

bryce-fitzsimons reviewed (dismissed):
we spoke over slack, got everything sorted out. PR looks good

lukesonnet commented on packages/back-end/src/services/experiments.ts:
```diff
@@ -823,6 +826,115 @@ export function toSnapshotApiInterface(
+    conditions: conditionsForDataSourceType(),
+
+    conversionDelayHours: behavior?.conversionWindowStart ?? 0,
+    conversionWindowHours: behavior?.conversionWindowEnd ?? 0,
```
This needs to be `conversionWindowHours:  behavior?.conversionWindowEnd - behavior?.conversionWindowStart`

tinahollygb replied:
Since both of those fields are optional, what do you think about one of these options:

With default conversion window (72):

```ts
    conversionWindowHours:
      (behavior?.conversionWindowEnd || DEFAULT_CONVERSION_WINDOW_HOURS) -
      (behavior?.conversionWindowStart ?? 0),
```

Or with 0 fallback:

```ts
    conversionWindowHours:
      (behavior?.conversionWindowEnd || 0) -
      (behavior?.conversionWindowStart ?? 0),
```

lukesonnet reviewed (commented):
Updated with some comments for conversion windows.

tinahollygb commented on packages/back-end/test/services/experiments.test.ts:
```diff
@@ -0,0 +1,310 @@
+          },
+          behavior: {
+            goal: "decrease",
+            conversionWindowStart: 10,
+            conversionWindowEnd: 50,
```
These result in:

```ts
        expect(result.conversionWindowHours).toEqual(40);
        expect(result.conversionDelayHours).toEqual(10);
```

From my understanding, this calculation is correct. Let me know if it isn't.

lukesonnet replied:
Yes, this is correct.

tinahollygb commented on packages/back-end/test/services/experiments.test.ts:
```diff
@@ -0,0 +1,310 @@
+        expect(result.userIdTypes).toEqual(["user_id"]);
+      });
+
+      it("with a full payload, should create a MetricInterface from a postMetric payload", () => {
```
This test is testing a full payload with the `sqlBuilder` property.

lukesonnet commented on packages/back-end/generated/spec.yaml:
```diff
@@ -664,6 +609,245 @@ paths:
+                    riskThresholdSuccess:
+                      type: number
+                    riskThresholdDanger:
+                      type: number
```
Let's add 

`riskThresholdDanger`: "Threshold for Risk to be considered too high, as a proportion (e.g. put 0.0125 for 1.25%)"
`riskThresholdSuccess`: "Threshold for Risk to be considered low enough, as a proportion (e.g. put 0.0025 for 0.25%)"

tinahollygb replied:
Addressed in [9af9ec6](https://github.com/growthbook/growthbook/pull/1116/commits/9af9ec620facaffde2bfdc68fdf60e02e7ee81ce)

lukesonnet reviewed (commented):
Some mostly minor suggestions about copy and validations in the comments.

I do have one question given your test example.

What happens if a user wants to POST a ratio metric and the metric they are going to use as the denominator in the same script? Right now, they have to:
1. Create the denominator
2. Use GET to get the denominator ID
3. Create the ratio metric, passing the above denominator ID

Is this the preferred flow? Is it easy to let users set the ID given this is a CREATE action after all?

tinahollygb commented on packages/back-end/generated/spec.yaml:
```diff
@@ -664,6 +609,245 @@ paths:
+                      type: number
+                    minPercentChange:
+                      type: number
+                      description: 'Minimum percent change to consider uplift significant, as a proportion (e.g. put 0.5 for 50%)'
```
Addressed in [9af9ec6](https://github.com/growthbook/growthbook/pull/1116/commits/9af9ec620facaffde2bfdc68fdf60e02e7ee81ce)

tinahollygb commented on packages/back-end/src/services/experiments.ts:
```diff
@@ -823,6 +827,222 @@ export function toSnapshotApiInterface(
+  // Validate conversion window
+  if (
+    typeof payload.behavior?.conversionWindowEnd !== "undefined" &&
+    typeof payload.behavior?.conversionWindowStart === "undefined"
```
Addressed in [9af9ec6](https://github.com/growthbook/growthbook/pull/1116/commits/9af9ec620facaffde2bfdc68fdf60e02e7ee81ce)

tinahollygb commented:
Regarding your question @lukesonnet:

> I do have one question given your test example.
> 
> What happens if a user wants to POST a ratio metric and the metric they are going to use as the denominator in the same script? Right now, they have to:
> 
> 1. Create the denominator
> 2. Use GET to get the denominator ID
> 3. Create the ratio metric, passing the above denominator ID

The create action should return the created metric. It will have the `id` property on it, which you can see in the response section. 

![CleanShot 2023-03-30 at 11 42 48@2x](https://user-images.githubusercontent.com/113377031/228933710-86e79077-38e0-4567-89fa-844740a8dbde.png)

So they'd need to create them sequentially, i.e. after creating the first one, get its `id` and use it when creating the second one. The developers should know how to do that.

> Is this the preferred flow? Is it easy to let users set the ID given this is a CREATE action after all?

While this is easy to allow, users should never be allowed to set their own `id`'s. These are unique identifiers and we cannot trust clients of the API to do that in a server-side application.

lukesonnet commented on packages/back-end/src/api/openapi/payload-schemas/PostMetricPayload.yaml:
```diff
@@ -0,0 +1,158 @@
+        type: number
+  sql:
+    type: object
+    description: "Preferred way to define SQL. Only one of `sql`, `sqlBuilder` or `mixpanel` allowed."
```
Sorry to add one more comment about copy, but maybe we could say the following for each of these three fields:

> Only one of `sql`, `sqlBuilder` or `mixpanel` allowed, and at least one must be specified.

lukesonnet reviewed (approved):
Understood re: the ID. Makes sense to me.

All looks good from the logic and field descriptions and requirement settings, although I did have one more suggestion to align the docs with the validatino. 

As usual, with this much `ts` I recommend someone else also take a look.

Also, there's kne problem with the PR description, but you probably just didn't update it. The `Orders metric in the API` request should fail now since it only has one of the conversion window fields. Minor, but just want to make sure that's because you didn't update it and not because for some reason this example request was valid.

tinahollygb commented:
Updated copy again @lukesonnet .

> The `Orders metric in the API` request should fail now since it only has one of the conversion window fields.

Updated it to add the end window—yes I have unit tests that test it would fail with only one.

lukesonnet reviewed (approved):
Thanks! Sorry for the iterative copy changes.

All good from my end.

bryce-fitzsimons commented on packages/back-end/src/services/experiments.ts:
```diff
@@ -806,6 +810,244 @@ export function toSnapshotApiInterface(
+        valid: false,
+        error: "riskThresholdSuccess must be a non-negative number",
+      };
+  }
```
we have this validation in the UI, not sure if you want to add it as well:
```
value.loseRisk < value.winRisk
      ? "The acceptable risk percentage cannot be higher than the too risky percentage"
      : "";
```

tinahollygb replied:
I've added this validation including tests.

bryce-fitzsimons commented on packages/back-end/test/services/experiments.test.ts:
```diff
@@ -0,0 +1,339 @@
+            conversionWindowStart: 10,
+            conversionWindowEnd: 50,
+            cap: 1337,
+            riskThresholdSuccess: 5,
```
might be a weird test case. both of the risk thresholds should be between 0 and 1

tinahollygb replied:
As discussed in the meeting, this feedback is invalid. Marking as resolved.

bryce-fitzsimons commented on packages/back-end/src/services/experiments.ts:
```diff
@@ -806,6 +810,244 @@ export function toSnapshotApiInterface(
+    if (
+      typeof behavior.riskThresholdSuccess !== "undefined" &&
+      behavior.riskThresholdSuccess < 0
+    )
```
should we check against `>1` as well? I can see people getting confused and entering "5" for 5% (which would actually be 500%)

tinahollygb replied:
I don't think this is correct—we can have float values, and the docs say:

> Threshold for Risk to be considered low enough, as a proportion (e.g. put 0.0025 for 0.25%). <br/> Must be a non-negative number and must not be higher than `riskThresholdDanger`

bryce-fitzsimons commented on packages/back-end/src/services/experiments.ts:
```diff
@@ -806,6 +810,244 @@ export function toSnapshotApiInterface(
+        error: "riskThresholdSuccess must be a non-negative number",
+      };
+  }
+
```
Should we have similar validations for `minPercentChange` and `maxPercentChange`? (`>=0`, `min <= max`)?

tinahollygb replied:
Added validation at the spec level

bryce-fitzsimons reviewed (changes requested):
a few small validation items to double check, but otherwise looks good.
